sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/ThroughputEstimatorTest.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction;
+
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assume.assumeThat;
+
+import com.google.cloud.Timestamp;
+import com.pholser.junit.quickcheck.From;
+import com.pholser.junit.quickcheck.Property;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.util.TimestampGenerator;
+
+public class ThroughputEstimatorTest {
+  private static final double DELTA = 1e-10;
+
+  @Property
+  public void testThroughputCalculation(@From(TimestampGenerator.class) Timestamp startTimestamp) {
```

Comment:
I think this specific test does not need to be a property based case, but we could create other test cases, such as:

1. Generate random pairs of timestamp / bytes within a 60 seconds window and verify if the bytes are correct
2. Generate random pairs of timestamp / bytes within a 300 seconds window and verify the sum at boundaries

Here is an example of how I would do #1:

```java
@Property
  public void testThroughputIsAccumulatedWithin60SecondsWindow(
      List<@InRange(minLong=0L, maxLong = 60000000L) Long> micros,
      List<Double> bytes
  ) {
    assumeThat(micros.size(), equalTo(bytes.size()));
    micros.sort(Long::compareTo);
    final List<Timestamp> timestamps = micros
        .stream()
        .map(Timestamp::ofTimeMicroseconds)
        .collect(Collectors.toList());
    final Double expectedSum = bytes.stream().reduce(Double::sum).get();

    for (int i = 0; i < timestamps.size(); i++) {
      estimator.update(timestamps.get(i), bytes.get(i))
    }
    double actualSum = estimator.getFrom(Timestamp
        .ofTimeSecondsAndNanos(timestamps.get(0).getSeconds() + 1, timestamps.get(0).getNanos()))
    assertEquals(expectedSum, actualSum, DELTA);
  }
```

Reply:
Generating two lists causes a random issue that the sizes of two lists sometimes do not match. So I decided to implement my own custom function to generate test data.